ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/forkchoice/ForkChoice.java

```diff
@@ -61,38 +63,42 @@ public void processHead(UInt64 nodeSlot) {
-                        final ForkChoiceStrategy forkChoiceStrategy = getForkChoiceStrategy();
-                        Bytes32 headBlockRoot =
-                            forkChoiceStrategy.findHead(
-                                transaction,
-                                finalizedCheckpoint,
-                                justifiedCheckpoint,
-                                justifiedCheckpointState.orElseThrow());
-
-                        recentChainData.updateHead(
-                            headBlockRoot,
-                            nodeSlot.orElse(
-                                forkChoiceStrategy
-                                    .blockSlot(headBlockRoot)
-                                    .orElseThrow(
-                                        () ->
-                                            new IllegalStateException(
-                                                "Unable to retrieve the slot of fork choice head"))));
-                        return transaction.commit();
-                      });
-            })
+    final Checkpoint retrievedJustifiedCheckpoint =
+        recentChainData.getStore().getJustifiedCheckpoint();
+    recentChainData
+        .retrieveCheckpointState(retrievedJustifiedCheckpoint)
+        .thenCompose(
+            justifiedCheckpointState ->
+                withLock(
+                    () -> {
+                      final Checkpoint finalizedCheckpoint =
+                          recentChainData.getStore().getFinalizedCheckpoint();
+                      final Checkpoint justifiedCheckpoint =
+                          recentChainData.getStore().getJustifiedCheckpoint();
+                      if (!justifiedCheckpoint.equals(retrievedJustifiedCheckpoint)) {
+                        LOG.info(
+                            "Skipping head block update as justified checkpoint was updated while loading checkpoint state.");
+                      }
+                      final StoreTransaction transaction = recentChainData.startStoreTransaction();
+                      final ForkChoiceStrategy forkChoiceStrategy = getForkChoiceStrategy();
+                      Bytes32 headBlockRoot =
+                          forkChoiceStrategy.findHead(
+                              transaction,
+                              finalizedCheckpoint,
+                              justifiedCheckpoint,
+                              justifiedCheckpointState.orElseThrow());
+
+                      recentChainData.updateHead(
+                          headBlockRoot,
+                          nodeSlot.orElse(
+                              forkChoiceStrategy
+                                  .blockSlot(headBlockRoot)
+                                  .orElseThrow(
+                                      () ->
+                                          new IllegalStateException(
+                                              "Unable to retrieve the slot of fork choice head"))));
```

Comment:
did we want to include the headBlockRoot in here?

Reply:
Done.  Pretty sure this is an "impossible" situation since we just found the root from protoarray and we're still holding the lock so nothing can change it in the meantime.